asapdiscovery-workflows/asapdiscovery/workflows/docking_workflows/ligand_transfer_docking.py

```diff
@@ -0,0 +1,548 @@
+
+    logger.info(f"Running ligand-transfer docking with inputs: {inputs}")
+    logger.info(f"Dumping input schema to {output_dir / 'inputs.json'}")
+    inputs.to_json_file(output_dir / "ligand_transfer_docking_inputs.json")
+
+    # make a directory to store intermediate CSV results
+    data_intermediates = Path(output_dir / "data_intermediates")
+    data_intermediates.mkdir(exist_ok=True)
+
+    if inputs.use_dask:
+        dask_client = make_dask_client_meta(
+            inputs.dask_type,
+            loglevel=inputs.loglevel,
+            n_workers=inputs.dask_n_workers,
+        )
+    else:
+        dask_client = None
+
+    # read structures
+    ref_complex_factory = MetaStructureFactory(
+        structure_dir=inputs.reference_complex_dir,
+        fragalysis_dir=inputs.reference_fragalysis_dir,
+        pdb_file=inputs.reference_pdb_file,
+    )
+    ref_complexes = ref_complex_factory.load(
+        use_dask=inputs.use_dask,
+        failure_mode=inputs.failure_mode,
+        dask_client=dask_client,
+    )
+
+    # read structures
+    target_factory = MetaStructureFactory(
+        structure_dir=inputs.target_structure_dir,
+        fragalysis_dir=inputs.target_fragalysis_dir,
+        pdb_file=inputs.target_pdb_file,
+    )
+    targets = target_factory.load(
+        use_dask=inputs.use_dask,
+        failure_mode=inputs.failure_mode,
+        dask_client=dask_client,
+    )
+
+    n_ref_complexes = len(ref_complexes)
+    logger.info(f"Loaded {n_ref_complexes} reference complexes")
+    n_targets = len(targets)
+    logger.info(f"Loaded {n_targets} target complexes")
+
+    # prep complexes
+    logger.info("Prepping complexes")
+    prepper = LigandTransferProteinPrepper(
+        reference_complexes=ref_complexes,
+        ref_chain=inputs.ref_chain,
+        active_site_chain=inputs.active_site_chain,
+        seqres_yaml=inputs.seqres_yaml,
+        loop_db=inputs.loop_db,
+    )
+    prepped_complexes = prepper.prep(
+        targets,
+        use_dask=inputs.use_dask,
+        dask_client=dask_client,
+        failure_mode=inputs.failure_mode,
+        cache_dir=inputs.cache_dir,
+        use_only_cache=inputs.use_only_cache,
+    )
+    del targets
+
+    n_prepped_complexes = len(prepped_complexes)
+    logger.info(f"Prepped {n_prepped_complexes} complexes")
+    for pc in prepped_complexes:
+        logger.info(f"Complex is {pc.target.target_name}")
+
+    if inputs.save_to_cache and inputs.cache_dir is not None:
+        logger.info(f"Writing prepped complexes to global cache {inputs.cache_dir}")
+        prepper.cache(prepped_complexes, inputs.cache_dir)
+
+    # Here the only thing that makes sense is the self docking selector
+    selector = StructureSelector.SELF_DOCKING.selector_cls()
+
+    # If the same reference is used for multiple targets, we will have duplicated ligands. This give problems when calling df.pivot()
+    def get_unique_ligands(complex_list):
```

Comment:
Had to add a function to check for repeated ligands, because the way the proteins are prepped right now is with a loop over the target x reference complexes, to assign each ligand from the reference to a target. So if we have multiple targets, we end up having a ligand repeated `n_targets` times, which gives problem when calling df.pivot() due to repeated entries. 
Not sure where is the best place to put this function, should I live it there?

Reply:
Yeah I am happy for it to live here, close to the call site makes it easy to understand.